crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,201 @@
+
+- `types.EllipsisType` (the sole inhabitant is `builtins.Ellipsis`).
+- `types.NotImplementedType` (the sole inhabitant is `builtins.NotImplemented`).
+- `None` (which can also be spelled as `Literal[None]`).
+    The sole inhabitant of the type is the runtime constant `None` itself.
+- `Literal[True]`: the sole inhabitant of the type is the constant `True`.
+- `Literal[False]`: the sole inhabitant of the type is the constant `False`.
+- `Literal[E.A]`, where `E` is an enum class which has a member `A` (the sole inhabitant is `E.A`).
+- A "literal class type": a type representing a single known class (and none of its possible subclasses).
+- A "literal function type": a type representing a single known function
+    (and excludes all other functions, even if they have the same signature).
+- A "literal module type": a type representing a single known module.
+
+Since it is known that all inhabitants of a given singleton type share the same memory address,
+singleton types in unions can be safely narrowed by identity,
+using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):
+
+```py
+def f(x: str | None):
+    if x is None:
+        ...  # x can be narrowed to `str`
+    else:
+        ...  # x can be narrowed to `None`
+```
+
+All singleton types are also sealed types, closed types and single-value types
+(see below for definitions).
+
+## Sealed types
+
+A sealed type is a type for which it is known
+that there is only a finite and pre-determined set of inhabitants at runtime.
+It follows from this that for every sealed type,
+there is also only a finite and pre-determined set of *subtypes* of that type.
+
+A sealed type in Python has the property
+that its proper subtypes[^2] are all disjunct from one another,
+and the union of all the proper subtypes of a sealed type is exactly equal to the sealed type.
+
+All singleton types are sealed types; however, not all sealed types are singleton types.
+
+Examples of sealed types (other than the singleton types listed above) are:
+
+- `bool`:
+
+    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
+    - The only proper subtypes of `bool` are `Literal[True]`, `Literal[False]`, and `Never`.
+        `Literal[True]`, `Literal[False]` and `Never` are all disjunct types,
+        and their union is exactly equal to `bool`.
+
+- Enums: consider the following enum class:
+
+    ```py
+    from enum import Enum, auto
+
+    class Foo(Enum):
+        X = auto()
+        Y = auto()
+        Z = auto()
+    ```
+
+    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
+        `Foo.X` and `Foo.Y`
+    - The only proper subtypes of `Foo` are `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]`, and `Never`.
+        `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]` and `Never` are all disjunct,
+        and the union `Literal[Foo.X, Foo.Y, Foo.Z] | Never` is exactly equal to the type `Foo`.
```

Comment:
This is not true, because it fails to consider unions of inhabitants, which are also proper subtypes of `Foo`. But I think if you just replace "proper subtype" with "inhabitant", leave out Never, and don't mention disjointness (since individual objects are by definition "disjoint" from each other), you can reach the same conclusion below (regarding how a negated intersection with the overall type is equivalent to a union of literal types for its inhabitants), which is the important part.